// ng-block-ui/bundles/ng-block-ui-router.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('ng-block-ui')) :
    typeof define === 'function' && define.amd ? define('ng-block-ui/router', ['exports', '@angular/core', 'ng-block-ui'], factory) :
    (factory((global['ng-block-ui'] = global['ng-block-ui'] || {}, global['ng-block-ui'].router = {}),global.ng.core,global['ng-block-ui']));
}(this, (function (exports,core,ngBlockUi) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var BlockUIPreventNavigation = (function () {
        function BlockUIPreventNavigation(blockUIService) {
            this.blockUIService = blockUIService;
        }
        /**
         * @return {?}
         */
        BlockUIPreventNavigation.prototype.canActivate = /**
         * @return {?}
         */
            function () {
                return !this.blockUIService.isActive(ngBlockUi.BLOCKUI_DEFAULT);
            };
        /**
         * @return {?}
         */
        BlockUIPreventNavigation.prototype.canActivateChild = /**
         * @return {?}
         */
            function () {
                return !this.blockUIService.isActive(ngBlockUi.BLOCKUI_DEFAULT);
            };
        BlockUIPreventNavigation.decorators = [
            { type: core.Injectable },
        ];
        /** @nocollapse */
        BlockUIPreventNavigation.ctorParameters = function () {
            return [
                { type: ngBlockUi.BlockUIService }
            ];
        };
        return BlockUIPreventNavigation;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */
    var BlockUIRouterModule = (function () {
        function BlockUIRouterModule() {
        }
        /**
         * @return {?}
         */
        BlockUIRouterModule.forRoot = /**
         * @return {?}
         */
            function () {
                return {
                    ngModule: BlockUIRouterModule,
                    providers: [
                        BlockUIPreventNavigation
                    ]
                };
            };
        BlockUIRouterModule.decorators = [
            { type: core.NgModule, args: [{
                        imports: [
                            ngBlockUi.BlockUIModule
                        ]
                    },] },
        ];
        return BlockUIRouterModule;
    }());

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes} checked by tsc
     */

    exports.BlockUIRouterModule = BlockUIRouterModule;
    exports.BlockUIPreventNavigation = BlockUIPreventNavigation;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
